Enable color displays

JorisVincent opened the pull request:
**Add support for color displays, on both regular GPU devices, as well as the ViewPixx3D.**
Specifically, it adds Graphics device classes for RGB-modes: `GPU_RGB` and `ViewPixx_RGB`. The original (greyscale) classes are renamed to `_grey`,  but aliased in the HRL code that calls for them. So, everything is backwards compatible: the tests of the public-facing HRL-class API pass without changes.

# Gamma-correction
For both the greyscale and the RGB mode, gamma correction is available. This requires passing in the Look-Up Table (LUT), or Color LUT. These have to be in the format specified in the `hrl.luts` module.

# Examples
This PR also adds examples in the documentation demonstrating the greyscale and RGB displays, in `docs/useful-examples/color/`. Each of the examples displays the same image, but in the different display modes. The `_raw_` examples use the raw greyscale or RGB data (no correction); the `_linear_` examples pass an example (C)LUT to "linearize" the values. On most displays, these examples should look more "washed out" than the raw examples.
Probably, these examples should be reorganized a bit to fit better into the documentation -- but they're there for now.

# Organization
This PR reorganizes the `Graphics` module and its objects, into three layers:
1. `Graphics` abstract base class (ABC) handles:
    - the OpenGL initialization and window management in its `.__init__()`
    - public-facing texture creation, through `.newTexture()`
    - public-facing setting of background, through `.changeBackground()`
 
2. Device specific concrete classes: `GPU_`, `ViewPixx_`,  `DataPixx`.
    - `GPU_` and `ViewPixx_` come in both `_grey` and `_RGB` variants. The `DataPixx` can only handle greyscale.
    - These classes handle converting an image array to the device-specific channels representation, which is then used to create a texture.
    - `DataPixx` and `ViewPixx_` classes also handle device initialization and connection to the hardware. This is now no longer the responsibility of the `HRL` class.
 
3. Mode-specific `Graphics_grey` and `Graphics_RGB` intermediate abstract base classes. These handle aspects that are mode specific,  but device independent, such as:
    - Validate dimensionality of the input image-arrays (HxW for greyscale; HxWx3 for RGB)
    - Gamma correction, using a (Color) Lookup Table
    - Implementation of mode-specific `.changeBackground()` using either a single value, or an RGB triplet.
    - In the future, this is also where mode-specific validation of the (C)LUT would take place.

guillermoaguilar committed:
commenting where changes need to be added

JorisVincent committed:
style(graphics): autoformat, clarifications

JorisVincent committed:
docs(color): example with RGB gamma correction

JorisVincent committed:
refactor(graphics.datapixx): update to match new organization

JorisVincent committed:
docs(color): update examples for new dispatching

JorisVincent committed:
docs(color): add examples for grey modes

JorisVincent committed:
docs(color): example for 'raw' RGB

JorisVincent committed:
refactor(graphics): separate classes for GPU grey, RGB

Separate classes define the conversion from image array to channels, and the gamma correction.
Dispatch to which object needs to be instantiated, based on ID string.
Also updates tests.

Closes #12

JorisVincent committed:
feat(graphics): separate `VIEWPixx` subclasses, for `_grey` and `_RGB`

Since VIEWPixx can do both. Also requires slightly different setup than DataPixx

JorisVincent committed:
refactor(graphics): move *Pixx hardware initialization into respective subclasses

JorisVincent committed:
refactor(graphics): remove outdated `channelsToInt`

Not called anywhere, so not necessary.

JorisVincent committed:
docs(graphics): unify docstrings

JorisVincent committed:
fix(graphics): add `.device` attribute to ABC

to hold bindings to hardware device. Defaults to `None`. Any subclasses can assign, if necessary (e.g., *Pixx devices)

JorisVincent committed:
refactor(graphics): add mode-specific (`grey`/`RGB`) intermediate ABCs

These contain some of the mode-specific logic:
- setting & validating the LUT vs CLUT
- setting the background (single value vs. triplet)
- validate image array dimensionality (single vs. 3-pane)

JorisVincent committed:
test(graphics): test for GPU_RGB device

JorisVincent committed:
fix(graphics): correctly deal with dimensionality of background

guillermoaguilar reviewed (commented):
Only one comment, otherwise it looks good to me

JorisVincent committed:
feat(hrl): add dynamic `hrl.device` property

for backwards compatibility, if any code relies on `my_hrl.device`. This just returns the `self.graphics.device`, because that it is actually what it should be doing.

guillermoaguilar commented on docs/useful-examples/color/example_lut.csv:
```diff
@@ -0,0 +1,257 @@
+intensity_in,intensity_out,luminance
```
let's maybe keep a list of the things that we know it will break. I don't know what is the good practice for keeping track of this.
